avogadro/qtplugins/labels/labels.cpp

```diff
@@ -0,0 +1,188 @@
+/******************************************************************************
+
+  This source file is part of the Avogadro project.
+
+  Copyright 2012 Kitware, Inc.
+
+  This source code is released under the New BSD License, (the "License").
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+******************************************************************************/
+
+#include "labels.h"
+
+#include <unordered_map>
+
+#include <avogadro/core/elements.h>
+#include <avogadro/core/molecule.h>
+#include <avogadro/qtgui/rwmolecule.h>
+#include <avogadro/rendering/cylindergeometry.h>
+#include <avogadro/rendering/geometrynode.h>
+#include <avogadro/rendering/groupnode.h>
+#include <avogadro/rendering/spheregeometry.h>
+#include <avogadro/rendering/textlabel3d.h>
+#include <avogadro/rendering/textproperties.h>
+
+#include <QtWidgets/QCheckBox>
+#include <QtWidgets/QDoubleSpinBox>
+#include <QtWidgets/QHBoxLayout>
+#include <QtWidgets/QLabel>
+#include <QtWidgets/QVBoxLayout>
+#include <QtWidgets/QWidget>
+
+namespace Avogadro {
+namespace QtPlugins {
+
+using Avogadro::QtGui::RWAtom;
+using Avogadro::QtGui::RWBond;
+using Avogadro::QtGui::RWMolecule;
+using Core::Elements;
+using Core::Molecule;
+using Rendering::GeometryNode;
+using Rendering::GroupNode;
+using Rendering::SphereGeometry;
+using Rendering::CylinderGeometry;
+
+typedef Avogadro::Core::Array<Molecule::BondType> NeighborListType;
+
+using namespace std;
+
+Labels::Labels(QObject* p)
+  : ScenePlugin(p), m_enabled(false), m_group(nullptr), m_setupWidget(nullptr),
+    m_multiBonds(true), m_showHydrogens(true)
+{
+}
+
+Labels::~Labels()
+{
+  if (m_setupWidget)
+    m_setupWidget->deleteLater();
+}
+
+void Labels::process(const Molecule& molecule, Rendering::GroupNode& node)
+{
+  Index hydrogenCount = 1;
+
+  m_group = &node;
+  GeometryNode* geometry = new GeometryNode;
+  node.addChild(geometry);
+
+  std::unordered_map<unsigned char, Index>
+    countMap; // Dictionary of atomic number and number of atoms
+
+  for (Index i = 0; i < molecule.atomCount(); ++i) {
+    Core::Atom atom = molecule.atom(i);
+    unsigned char atomicNumber = atom.atomicNumber();
+    if (atomicNumber == 1)
+      continue; // Do not process Hydrogens separately
+
+    std::unordered_map<unsigned char, Index>::const_iterator lookUp;
+    lookUp = countMap.find(atomicNumber);
+
+    if (lookUp == countMap.end())
+      countMap[atomicNumber] = 1; // For first occurrence of atom
+
+    else
+      countMap[atomicNumber] += 1; // For other occurrences of atom
+
+    if (!m_showHydrogens)
```

Comment:
This looks like nothing will be rendered if showHydrogens is false, which doesn't seem right.